File: Cpp_part/Leetcode/Leetcode/array/167. 两数之和 II - 输入有序数组/Solution2.cpp
#include <iostream>
#include <vector>
using namespace std;
/*双指针法
 双指针法本质上是 decrease a constant的方法。 只用该方法的前提在于，可以利用当前结果，定向的缩小范围。这里是移动指针
 */
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int left = 0, right = numbers.size() - 1;
        while (left < right ){
            int cur_sum = numbers[left] + numbers[right];
            if (cur_sum == target)
                return vector<int> {left+1, right+1};
            else if (cur_sum < target)
                left ++;
            else
                right --;
        };
        return vector<int>{-1, -1};
    }
};
//
//int main(){
//    Solution* s = new Solution;
//    vector<int> input = {2, 7, 15, 11};
//    vector<int> res = s -> twoSum(input, 9);
//}

File: python_part/Leetcode/others/Sliding window/面试题57 - II. 和为s的连续正数序列/Solution2.py
'''
杀鸡为何用牛刀？solution1 的解题方法太复杂了，我们只要根据target和现在window中的值来决定是移动left 还是right还是记录数据就行了，
因此我们每次只移动left or right 其中一个，并不在一个循环里执行多次操作
'''
from typing import List
class Solution:
    def findContinuousSequence(self, target: int) -> List[List[int]]:
        left = right = 1
        window = 0
        res = []
        while right <= int(target/2)+2:
            if target > window:
                window += right
                right += 1
            elif target < window:
                window -= left
                left += 1
            else:
                res.append(list(range(left, right)))
                window -= left
                left += 1
        return res
foo = Solution()
print(foo.findContinuousSequence(15))